walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,101 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Build a Collection of every possible attendee
+    Collection<String> allAttendees = new HashSet<>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+
+    // Determine possible meeting times for every possible attendee 
+    MeetingRequest requestEveryone = new MeetingRequest(allAttendees, request.getDuration());
+    Collection<TimeRange> queryEveryone = queryGivenAttendees(events, requestEveryone);
+		
+    if (!queryEveryone.isEmpty()) {
+      return queryEveryone;
+    }
+    else {
+      return queryGivenAttendees(events, request);
+    }
+  }
+
+  /** Returns the possible meeting times given required attendees */
+  public Collection<TimeRange> queryGivenAttendees(Collection<Event> events, MeetingRequest request) {
+    ArrayList<TimeRange> busy = new ArrayList<>();
+		
+    // Determine times where meeting cannot be scheduled
+    for (Event event : events) { 
+      if (overlappingAttendees(event.getAttendees(), request.getAttendees())) {
+        busy.add(event.getWhen());
+      }
+    }
+
+    // Sort these times by start time
+    Collections.sort(busy, TimeRange.ORDER_BY_START);
+
+    // Simplify the times to avoid overlap
+    ArrayList<TimeRange> mergedBusy = merge(busy); 
+
+    // Get the complement 
+    Collection<TimeRange> free = new ArrayList<>();
+    if (mergedBusy.size() > 0) { 
+      // Add time at start of day if possible
+      if (mergedBusy.get(0).start() - TimeRange.START_OF_DAY >= request.getDuration()) {
+        free.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, mergedBusy.get(0).start(), false));
+      }
+      // Fill in the time between unavailable times
+      for (int i = 0; i < mergedBusy.size() - 1; i++) {
+        if (mergedBusy.get(i + 1).start() - mergedBusy.get(i).end() >= request.getDuration()) {
+          free.add(TimeRange.fromStartEnd(mergedBusy.get(i).end(), mergedBusy.get(i + 1).start(), false));
+        }
+      }
+      // Add time at end of day if possible
+      if (TimeRange.END_OF_DAY - mergedBusy.get(mergedBusy.size() - 1).end() >= request.getDuration()) {
+        free.add(TimeRange.fromStartEnd(mergedBusy.get(mergedBusy.size() - 1).end(), TimeRange.END_OF_DAY, true));
+      }
+    }
+    // Add the whole day if possible
+    else if (TimeRange.WHOLE_DAY.duration() >= request.getDuration() && request.getAttendees().size() > 0) {
+      free.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+    }
+    return free;
```

Comment:
Consider moving some of the logic in the `if` conditions to methods, just to help make this more readable.

Reply:
Fixed - I moved this over to a new `complementOf` method